File: problem/Balance_Binary_Search_Tree/balanceBST.js
/**
 * @param {TreeNode} root
 * @return {TreeNode}
 */
var balanceBST = function(root) {
    const vals = [];
    const inorder = (node) => {
        if (!node) return;
        inorder(node.left);
        vals.push(node.val);
        inorder(node.right);
    };

    const build = (l, r) => {
        if (l > r) return null;
        const mid = Math.floor((l + r) / 2);
        const node = new TreeNode(vals[mid]);
        node.left = build(l, mid - 1);
        node.right = build(mid + 1, r);
        return node;
    };

    inorder(root);
    return build(0, vals.length - 1);
};

File: problem/Balance_Binary_Search_Tree/balanceBST.c
void inorder(struct TreeNode *node, int *vals, int *size)
{
    if (!node)
        return;
    inorder(node->left, vals, size);
    vals[(*size)++] = node->val;
    inorder(node->right, vals, size);
}

struct TreeNode *build(int *vals, int l, int r)
{
    if (l > r)
        return NULL;
    int mid = l + (r - l) / 2;
    struct TreeNode *node = (struct TreeNode *)malloc(sizeof(struct TreeNode));
    node->val = vals[mid];
    node->left = build(vals, l, mid - 1);
    node->right = build(vals, mid + 1, r);
    return node;
}

struct TreeNode *balanceBST(struct TreeNode *root)
{
    int *vals = (int *)malloc(10000 * sizeof(int));
    int size = 0;
    inorder(root, vals, &size);
    return build(vals, 0, size - 1);
}